### PR TITLE
Adding back spring security

### PR DESCRIPTION
**Pull request by kyleaolson**

### CODE REVIEW

File: apis/src/main/java/xyz/bookself/config/BookselfBasicAuthConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http.authorizeRequests()
+                .antMatchers("/**").permitAll() // Allow everything except above
+                .and()
+                .csrf().disable()
+                .httpBasic();
+    }
```
Comment (kyleaolson) [inline apis/src/main/java/xyz/bookself/config/BookselfBasicAuthConfiguration.java:47]: Permitting all -- this gets Spring Security ready for us to manually authorize on a method-by-method basis.

File: apis/src/main/java/xyz/bookself/controllers/user/UserController.java [hunk 2]
```diff
@@ -29,9 +26,22 @@ public UserController(UserRepository repository) {
...
+        // If somebody is logged in but is trying to access somebody else's profile, FORBIDDEN
+        if (!userDetails.getId().equals(id)) {
+            return new ResponseEntity<>(HttpStatus.FORBIDDEN);
+        }
+
+        // Otherwise look up the user
+        return userRepository.findById(id)
+                .map(u -> new ResponseEntity<>(u, HttpStatus.OK))
+                .orElseGet(() -> new ResponseEntity<>(HttpStatus.NOT_FOUND));
     }
```
Comment (kyleaolson) [inline apis/src/main/java/xyz/bookself/controllers/user/UserController.java:45]: @davideengel / @jaredkschreiber can you confirm this approach? The set up is to only allow a user to pull their own details when authenticated.

File: apis/src/main/resources/logback.xml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    <!-- Setup default Spring logging -->
+    <include resource="org/springframework/boot/logging/logback/base.xml"/>
+
+    <!-- Add papertrail logging -->
+    <appender name="PAPERTRAIL" class="ch.qos.logback.classic.net.SyslogAppender">
+        <syslogHost>logs.papertrailapp.com</syslogHost>
+        <port>28499</port>
+        <facility>USER</facility>
+        <suffixPattern>bookself-backend: %logger %msg</suffixPattern>
+    </appender>
```
Comment (kyleaolson) [inline apis/src/main/resources/logback.xml:12]: Also setting up logging aggregation with Papertrail. Will get accounts sent out later.

File: apis/src/test/java/xyz/bookself/security/WithBookselfUserDetails.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+import org.springframework.security.test.context.support.WithSecurityContext;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+@WithSecurityContext(factory = WithBookselfUserDetailsSecurityContextFactory.class)
+public @interface WithBookselfUserDetails {
+    int id();
+}
```
Comment (kyleaolson) [inline apis/src/test/java/xyz/bookself/security/WithBookselfUserDetails.java:12]: Custom annotation allows us to add mocked authentication on a test case level

File: user-service/src/main/java/xyz/bookself/users/repository/UserRepository.java [hunk 5]
```diff
@@ -4,9 +4,12 @@
...
 import xyz.bookself.users.domain.User;
 
-import java.util.Collection;
+import java.util.Optional;
 
 public interface UserRepository extends JpaRepository<User, Integer> {
-    @Query(nativeQuery = true, value = "SELECT * FROM users WHERE email = ?1")
-    User findUserByEmail(String email);
+    @Query(nativeQuery = true, value = "SELECT * FROM users WHERE LOWER(email)=LOWER(?1)")
+    Optional<User> findUserByEmail(String email);
```
Comment (kyleaolson) [inline user-service/src/main/java/xyz/bookself/users/repository/UserRepository.java:11]: Updated this to return an Optional to be consistent.

**Review by nardost — APPROVED** [review]

**Pull request merged**